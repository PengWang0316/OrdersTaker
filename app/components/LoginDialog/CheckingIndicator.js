import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { CircularProgress, Tooltip } from '@material-ui/core';
import { CheckCircle, Error } from '@material-ui/icons';
import { withStyles } from '@material-ui/core/styles';
import { orange, red, green } from '@material-ui/core/colors';

const adjustPosition = {
  position: 'relative',
  top: -10
};

const styles = {
  circularProgress: {
    ...adjustPosition,
    color: orange[800]
  },
  checkCircle: {
    ...adjustPosition,
    color: green[500]
  },
  error: {
    ...adjustPosition,
    color: red[500]
  }
};

export const CheckingIndicator = ({ classes, isChecking, isAvaliable }) => (
  <Fragment>
    {isChecking && <CircularProgress className={classes.circularProgress} size={20} />}
    {!isChecking && (isAvaliable ? <CheckCircle className={classes.checkCircle} size={20} /> : (
      <Tooltip id="usernameTooltip" title="This user name has already been taken.">
        <Error className={classes.error} size={20} />
      </Tooltip>
    ))}
  </Fragment>
);
CheckingIndicator.propTypes = {
  classes: PropTypes.object.isRequired,
  isChecking: PropTypes.bool.isRequired,
  isAvaliable: PropTypes.bool.isRequired
};
export default withStyles(styles)(CheckingIndicator);
