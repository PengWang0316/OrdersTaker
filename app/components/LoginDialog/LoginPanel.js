import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core/styles';
import { Typography, Button, TextField } from '@material-ui/core';

import facebookLoginButton from '../../images/FacebookLogin.png';
import googleLoginButton from '../../images/GoogleLogin.png';
import { API_FACEBOOK_LOGIN, API_GOOGLE_LOGIN } from '../../actions/ApiUrls';
import { loginWithPassword, emptyUser } from '../../actions/UserActions';

const LOGIN_FAIL_MESSAGE = 'Wrong username or password';

/* istanbul ignore next */
const getDividerBeforAfter = color => ({ // Generating a basic styling object for divider's before and after.
  borderTop: '1px solid',
  borderTopColor: color,
  content: '""',
  height: 1,
  top: '50%',
  width: '45%',
  position: 'absolute'
});
/* istanbul ignore next */
const styles = theme => ({
  title: {
    fontWeight: 'bold',
    fontSize: 16,
    marginBottom: 15
  },
  fontWeight700: {
    fontWeight: 700
  },
  loginBtn: {
    width: '100%',
    maxWidth: 220,
    marginBottom: 15,
    '&:hover': {
      cursor: 'pointer'
    }
  },
  divider: {
    margin: '20px 0',
    position: 'relative',
    textAlign: 'center',
    '&:before': {
      ...getDividerBeforAfter(theme.palette.primary.light),
      left: -6
    },
    '&:after': {
      ...getDividerBeforAfter(theme.palette.primary.light),
      right: -6
    }
  },
  textField: {
    width: '100%'
  },
  loginButtonDiv: {
    textAlign: 'right'
  },
  regesterBtn: {
    fontSize: 12,
    textTransform: 'capitalize'
  }
});

/**
 * The componnet that shows users login options.
 */
export class LoginPanel extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    onTogglePanels: PropTypes.func.isRequired,
    onToggleSnackbar: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    loginWithPassword: PropTypes.func.isRequired,
    emptyUser: PropTypes.func.isRequired
  };

  state = {
    username: '',
    password: '',
    isReady: false, // The indicator of login button.
    isSubmitted: false // After a user submitted a login form, change it to true in order to show the error snackbar. It can prevent the snackbar shows in the first time a user open the dialog.
  };

  /**
   * Call the onToggleSnackbar when isSubmitted state is true and props received a null for user.
   * @param {object} nextProps contains the new props value.
   * @param {*} prevState contains the previous state's value.
   * @return {Object} Return an object to change the state.
   */
  static getDerivedStateFromProps(nextProps, prevState) {
    if (prevState.isSubmitted && nextProps.user.isFail) {
      nextProps.onToggleSnackbar(LOGIN_FAIL_MESSAGE);
      nextProps.emptyUser(); // Reset user state to an empty object. So, next time the back-end code still can send a isFail mark when the authentication fails.
      return { isSubmitted: false };
    }
    return null;
  }

  /**
   * Set state when a user type in the text fields.
   * @param {object} event is the html element the user is interacting with.
   * @return {null} No return.
   */
  handleTextChanged = ({ target }) => this.setState({
    [target.id]: target.value
  }, () => this.setState({
    isReady: !!(this.state.username && this.state.password)
  }));

  /**
   * Call the login action when a user clicks the login button.
   * @return {null} No return.
   */
  handlLoginButtonClick = () => this.setState(
    { isReady: false, isSubmitted: true },
    () => this.props.loginWithPassword({
      username: this.state.username, password: this.state.password
    })
  );

  /**
   * The render method.
   * @return {jsx} Return jsx.
   */
  render() {
    const { classes, onTogglePanels } = this.props;
    const { isReady, username, password } = this.state;
    return (
      <Fragment>
        <Typography color="primary" className={classes.title} data-testid="loginPanelTitle">Social Login</Typography>
        <div>
          <a href={API_FACEBOOK_LOGIN}>
            <img className={classes.loginBtn} src={facebookLoginButton} alt="Use Facebook account login" />
          </a>
        </div>
        <div>
          <a href={API_GOOGLE_LOGIN}>
            <img className={classes.loginBtn} src={googleLoginButton} alt="Use Google account login" />
          </a>
        </div>

        <div className={classes.divider}>
          <Typography color="primary" className={classes.fontWeight700}>or</Typography>
        </div>

        <Typography color="primary" className={classes.title}>UserName Login</Typography>

        <form noValidate autoComplete="off">
          <div>
            <TextField
              autoFocus
              id="username"
              label="Username"
              className={classes.textField}
              margin="normal"
              value={username}
              onChange={this.handleTextChanged}
            />
          </div>
          <div>
            <TextField
              id="password"
              label="Password"
              className={classes.textField}
              type="password"
              autoComplete="current-password"
              margin="normal"
              value={password}
              onChange={this.handleTextChanged}
            />
          </div>
          <div className={classes.loginButtonDiv}>
            <Button size="small" aria-label="Login" color="primary" disabled={!isReady} onClick={this.handlLoginButtonClick}>Login</Button>
          </div>
        </form>
        <div className={classes.loginButtonDiv}>
          <Button onClick={onTogglePanels} className={classes.regesterBtn} size="small" aria-label="Login" color="primary" data-testid="switchRegisterButton">Do not have an account? Register one</Button>
        </div>
      </Fragment>
    );
  }
}
/* istanbul ignore next */
const mapStateToProps = state => ({ user: state.user });
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  loginWithPassword: user => dispatch(loginWithPassword(user)),
  emptyUser: () => dispatch(emptyUser())
});
export default connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(LoginPanel));
