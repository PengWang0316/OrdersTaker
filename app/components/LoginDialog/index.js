// Setting up a default import from this folder.
import LoginDialog from './LoginDialog';

export default LoginDialog;
