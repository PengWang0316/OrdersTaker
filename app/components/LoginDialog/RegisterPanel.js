import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { TextField, Typography, Button, Tooltip, CircularProgress } from '@material-ui/core';
import green from '@material-ui/core/colors/green';
import { withStyles } from '@material-ui/core/styles';
import { Check, Search, CheckCircle, Error } from '@material-ui/icons';

import { registerUser, checkUsernameAvailable } from '../../actions/UserActions';
import CheckingIndicator from './CheckingIndicator';

const styles = {
  title: {
    fontSize: 16,
    fontWeight: 'bold'
  },
  registerBtnDiv: {
    textAlign: 'right',
    margin: '20px 0'
  },
  checkIcon: {
    backgroundColor: green[500],
    color: 'white',
    marginRight: 20,
    '&:hover': {
      backgroundColor: green[700]
    }
  },
  circularProgress: {
    color: green[500],
    position: 'relative',
    right: -49,
    top: 25
  },
  switchBtn: {
    textTransform: 'capitalize'
  },
  usernameDiv: {
    display: 'flex',
    alignItems: 'flex-end'
  }
  // '@media (max-width: 450px)': {
  //   textField: {
  //     width: '100%'
  //   }
  // }
};

const USERNAME_REGEXP = /[^\w\d_@]/g;
const EMAIL_REGEXP = /^[\w\d]+@.+\..+/g;
const USERNAME_TIP_MESSAGE = 'Characters, number, _ and @';
const PASSWORD_TIP_MESSAGE = 'Passwords should be same';
const EMAIL_TIP_MESSAGE = 'Wrong email address';
/**
 * Show the register panel.
 */
export class RegisterPanel extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    onTogglePanels: PropTypes.func.isRequired,
    onToggleSnackbar: PropTypes.func.isRequired,
    registerUser: PropTypes.func.isRequired
  };

  state = {
    username: '',
    password: '',
    repeatPassword: '',
    email: '',
    isReady: false, // The indicator that is used to disable register button.
    isUsernameError: false,
    isPasswordError: false,
    isEmailError: false,
    isNameAvalible: false, // whether username is avalible to be used.
    isChecking: false, // whether checking the avalibility for the user name.
    isWaiting: false // When register button is hit turn this to true.
  };

  /**
   * TODO may have a better way to make it as a libaray for the future reusage.
   * Setting the state value to the text field's value when a user changed it.
   * @param {object} event is the click event.
   * @return {null} No return.
   */
  handleTextFieldChange = ({ target }) => {
    const { onToggleSnackbar } = this.props;
    this.setState({ [target.id]: target.value }, () => {
      // If all field is not empty, set the isReady to true.
      this.setState(({
        username, password, repeatPassword, email
      }) => ({ isReady: !!(username && password && repeatPassword && email) }));

      // Do the validation
      switch (target.id) {
        case 'username':
          this.setState({ isUsernameError: false, isReady: false }); // Erasing the error state and check it again. If username is changed, alway set the isReady to false and let the checking function callback to decide it state.
          if (target.value.match(USERNAME_REGEXP)) {
            onToggleSnackbar(USERNAME_TIP_MESSAGE);
            this.setState({ isUsernameError: true });
          } else {
            // Setting up a timeout function to check the avalibility for the username.
            if (this.checkUsernameTimeout) clearTimeout(this.checkUsernameTimeout); // If the check function has already been schulded, clear the old one.
            this.checkUsernameTimeout = setTimeout(() => {
              this.setState({ isChecking: true });
              checkUsernameAvailable(target.value).then(isAvalible => this.setState(({
                username, password, repeatPassword, email
              }) => ({ isReady: !!(username && password && repeatPassword && email && !!isAvalible), isChecking: false, isNameAvalible: !!isAvalible })));
            }, 1000);
          }
          break;
        case 'password':
          this.setState({ isPasswordError: false }); // Erasing the error state and check it again.
          if (target.value !== this.state.repeatPassword) {
            onToggleSnackbar(PASSWORD_TIP_MESSAGE);
            this.setState({ isPasswordError: true, isReady: false });
          }
          break;
        case 'repeatPassword':
          this.setState({ isPasswordError: false }); // Erasing the error state and check it again.
          if (target.value !== this.state.password) {
            onToggleSnackbar(PASSWORD_TIP_MESSAGE);
            this.setState({ isPasswordError: true, isReady: false });
          }
          break;
        case 'email':
          this.setState({ isEmailError: false }); // Erasing the error state and check it again.
          if (!target.value.match(EMAIL_REGEXP)) {
            onToggleSnackbar(EMAIL_TIP_MESSAGE);
            this.setState({ isEmailError: true, isReady: false });
          }
          break;
        default:
          break;
      }
    });
  };

  /**
   * Disable the register button, set isWaiting to true, and call the registerUser function in UserActions.
   * @return {null} No return.
   */
  handleRegisterBtnClick = () => {
    this.setState({
      isReady: false,
      isWaiting: true
    });
    this.props.registerUser({
      username: this.state.username,
      password: this.state.password,
      email: this.state.email
    });
  };

  /**
   * Rendering the jsx.
   * @return {jsx} Return jsx.
   */
  render() {
    const { classes, onTogglePanels } = this.props;
    const {
      username, password, repeatPassword, email,
      isUsernameError, isPasswordError, isEmailError, isReady, isWaiting, isChecking, isNameAvalible
    } = this.state;
    return (
      <Fragment>
        <Typography className={classes.title} color="primary" data-testid="registerPanelTitle">Register a new user</Typography>
        <form noValidate autoComplete="off">
          <div className={classes.usernameDiv}>
            <Tooltip id="tooltipUsername" title={USERNAME_TIP_MESSAGE}>
              <TextField
                error={isUsernameError}
                helperText={isUsernameError ? USERNAME_TIP_MESSAGE : null}
                autoFocus
                id="username"
                label="Username"
                margin="normal"
                type="email"
                value={username}
                onChange={this.handleTextFieldChange}
                disabled={isWaiting}
                fullWidth
              />
            </Tooltip>
            {username !== '' && <CheckingIndicator isChecking={isChecking} isAvaliable={isNameAvalible} />}
          </div>
          <div>
            <TextField
              id="password"
              label="Password"
              margin="normal"
              type="password"
              value={password}
              onChange={this.handleTextFieldChange}
              disabled={isWaiting}
              fullWidth
            />
          </div>
          <Tooltip id="tooltipRpPw" title="Please type in your password again">
            <TextField
              error={isPasswordError}
              helperText={isPasswordError ? PASSWORD_TIP_MESSAGE : null}
              id="repeatPassword"
              label="Repeat Password"
              margin="normal"
              type="password"
              value={repeatPassword}
              onChange={this.handleTextFieldChange}
              disabled={isWaiting}
              fullWidth
            />
          </Tooltip>
          <div>
            <TextField
              error={isEmailError}
              helperText={isEmailError ? EMAIL_TIP_MESSAGE : null}
              id="email"
              label="Email"
              margin="normal"
              type="email"
              value={email}
              onChange={this.handleTextFieldChange}
              disabled={isWaiting}
              fullWidth
            />
          </div>
          <Tooltip placement="bottom-end" id="tooltipRegisterBtn" title={isReady ? 'Click to register' : 'Please fill out all field above'}>
            <div className={classes.registerBtnDiv}>
              {isWaiting && (
                <Fragment>
                  <CircularProgress size={58} className={classes.circularProgress} />
                  <Button size="medium" mini className={classes.checkIcon} variant="fab"><Check /></Button>
                </Fragment>
              )}
              <Button onClick={this.handleRegisterBtnClick} color="primary" aria-label="Register" size="small" disabled={!isReady} variant="contained">Register</Button>
            </div>
          </Tooltip>
          <div>
            <Button onClick={onTogglePanels} color="primary" className={classes.switchBtn} data-testid="switchLoginButton">Have an account already? Go to login</Button>
          </div>
        </form>

      </Fragment>
    );
  }
}
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  registerUser: user => dispatch(registerUser(user))
});
export default connect(null, mapDispatchToProps)(withStyles(styles)(RegisterPanel));
