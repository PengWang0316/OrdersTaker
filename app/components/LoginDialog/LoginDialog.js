import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Dialog, DialogContent, IconButton } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import { Close } from '@material-ui/icons/';

import dialogStyles from '../../styles/DialogSytels';
import LoginPanel from './LoginPanel';
import RegisterPanel from './RegisterPanel';

/* istanbul ignore next */
const styles = theme => ({
  ...dialogStyles, // Extending some basic dialog styles from DialogSytels.
});

/**
 * Showing the login and register panels in one dialog.
 */
export class LoginDialog extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    open: PropTypes.bool.isRequired,
    onClose: PropTypes.func.isRequired,
    onToggleSnackbar: PropTypes.func.isRequired
  };

  state = {
    isLogin: true // Control whether to show the login or register panel.
  };

  /**
   * Switch the isLogin state to an opposite value.
   * @return {null} No return.
   */
  togglePanels = () => this.setState(({ isLogin }) => ({ isLogin: !isLogin }));

  /**
   * Rendering method.
   * @return {jsx} Return jsx.
   */
  render() {
    const {
      classes, open, onClose, onToggleSnackbar
    } = this.props;
    const { isLogin } = this.state;
    return (
      <Dialog onClose={onClose} open={open} classes={{ paper: classes.dialogPaper }}>
        <DialogContent className={classes.dialogContent}>
          <IconButton onClick={onClose} className={classes.dialogCloseButton}><Close className={classes.dialogCloseIcon} /></IconButton>
          {isLogin ? <LoginPanel onTogglePanels={this.togglePanels} onToggleSnackbar={onToggleSnackbar} /> : <RegisterPanel onTogglePanels={this.togglePanels} onToggleSnackbar={onToggleSnackbar} />}
        </DialogContent>
      </Dialog>
    );
  }
}
export default withStyles(styles)(LoginDialog);
