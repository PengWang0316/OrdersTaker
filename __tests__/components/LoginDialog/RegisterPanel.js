import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { RegisterPanel } from '../../../app/components/LoginDialog/RegisterPanel';

jest.mock('@material-ui/core/TextField', () => 'TextField');
jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/core/Button', () => 'Button');
jest.mock('@material-ui/core/Tooltip', () => 'Tooltip');
jest.mock('@material-ui/core/CircularProgress', () => 'CircularProgress');
jest.mock('@material-ui/icons/Check', () => 'Check');
jest.mock('../../../app/components/LoginDialog/CheckingIndicator', () => 'CheckingIndicator');
jest.mock('../../../app/actions/UserActions', () => ({ checkUsernameAvailable: jest.fn().mockReturnValue(Promise.resolve('true')) }));

console.error = jest.fn();

describe('RegisterPanel', () => {
  const defaultProps = {
    classes: {
      title: 'title',
      registerBtnDiv: 'registerBtnDiv',
      checkIcon: 'checkIcon',
      circularProgress: 'circularProgress',
      switchBtn: 'switchBtn'
    },
    onTogglePanels: jest.fn(),
    onToggleSnackbar: jest.fn(),
    registerUser: jest.fn()
  };
  const getShallowComponent = (props = defaultProps) => shallow(<RegisterPanel {...props} />);

  test('Initial states', () => {
    const component = getShallowComponent();
    const {
      username, password, repeatPassword, email, isReady, isUsernameError, isPasswordError, isEmailError, isWaiting, isNameAvalible, isChecking
    } = component.state();
    expect(username).toBe('');
    expect(password).toBe('');
    expect(repeatPassword).toBe('');
    expect(email).toBe('');
    expect(isReady).toBe(false);
    expect(isUsernameError).toBe(false);
    expect(isPasswordError).toBe(false);
    expect(isEmailError).toBe(false);
    expect(isWaiting).toBe(false);
    expect(isNameAvalible).toBe(false);
    expect(isChecking).toBe(false);
  });

  test('handleTextFieldChange', () => {
    const UserActions = require('../../../app/actions/UserActions');
    const component = getShallowComponent();

    jest.useFakeTimers();
    component.instance().handleTextFieldChange({ target: { id: 'username', value: 'V8al_u@e' } });
    expect(component.state('username')).toBe('V8al_u@e');
    expect(component.state('isReady')).toBe(false);
    expect(component.state('isUsernameError')).toBe(false);
    expect(defaultProps.onToggleSnackbar).not.toHaveBeenCalled();
    expect(clearTimeout).not.toHaveBeenCalled();
    expect(setTimeout).toHaveBeenCalledTimes(1);
    // component.setState({ isReady: true }); // Manually set the isReady to true in order to test it later.
    jest.runAllTimers();
    expect(component.state('isChecking')).toBe(true);
    expect(UserActions.checkUsernameAvailable).toHaveBeenCalledTimes(1);
    expect(UserActions.checkUsernameAvailable).toHaveBeenLastCalledWith('V8al_u@e');
    // expect(component.state('isReady')).toBe(false);
    jest.useFakeTimers();
    component.instance().checkUsernameTimeout = {};
    component.instance().handleTextFieldChange({ target: { id: 'username', value: 'V8al_u@e' } });
    expect(clearTimeout).toHaveBeenCalledTimes(1);

    component.instance().handleTextFieldChange({ target: { id: 'username', value: 'va!lue' } });
    expect(component.state('isReady')).toBe(false);
    expect(component.state('isUsernameError')).toBe(true);
    expect(defaultProps.onToggleSnackbar).toHaveBeenCalledTimes(1);
    expect(defaultProps.onToggleSnackbar).toHaveBeenLastCalledWith('Characters, number, _ and @');
    // component.instance().handleTextFieldChange({ target: { id: 'username', value: 'V8al_u@e' } });
    component.setState({ username: 'V8al_u@e' });

    component.instance().handleTextFieldChange({ target: { id: 'password', value: 'password' } });
    expect(component.state('password')).toBe('password');
    expect(component.state('isPasswordError')).toBe(true);
    expect(component.state('isReady')).toBe(false);
    expect(defaultProps.onToggleSnackbar).toHaveBeenCalledTimes(2);
    expect(defaultProps.onToggleSnackbar).toHaveBeenLastCalledWith('Passwords should be same');
    component.setState({ repeatPassword: 'password' });
    component.instance().handleTextFieldChange({ target: { id: 'password', value: 'password' } });
    expect(component.state('isPasswordError')).toBe(false);
    expect(component.state('isReady')).toBe(false);
    expect(defaultProps.onToggleSnackbar).toHaveBeenCalledTimes(2);

    component.instance().handleTextFieldChange({ target: { id: 'repeatPassword', value: 'password' } });
    expect(component.state('repeatPassword')).toBe('password');
    expect(component.state('isReady')).toBe(false);
    expect(component.state('isPasswordError')).toBe(false);
    expect(defaultProps.onToggleSnackbar).toHaveBeenCalledTimes(2);
    component.instance().handleTextFieldChange({ target: { id: 'repeatPassword', value: 'password1' } });
    expect(component.state('repeatPassword')).toBe('password1');
    expect(component.state('isPasswordError')).toBe(true);
    expect(component.state('isReady')).toBe(false);
    expect(defaultProps.onToggleSnackbar).toHaveBeenCalledTimes(3);
    expect(defaultProps.onToggleSnackbar).toHaveBeenLastCalledWith('Passwords should be same');
    component.instance().handleTextFieldChange({ target: { id: 'repeatPassword', value: 'password' } });

    component.instance().handleTextFieldChange({ target: { id: 'email', value: 'd@d.c' } });
    expect(component.state('email')).toBe('d@d.c');
    expect(component.state('isReady')).toBe(true);
    expect(component.state('isEmailError')).toBe(false);
    expect(defaultProps.onToggleSnackbar).toHaveBeenCalledTimes(3);
    component.instance().handleTextFieldChange({ target: { id: 'email', value: 'dd.c' } });
    expect(component.state('isReady')).toBe(false);
    expect(component.state('isEmailError')).toBe(true);
    expect(defaultProps.onToggleSnackbar).toHaveBeenCalledTimes(4);
    expect(defaultProps.onToggleSnackbar).toHaveBeenLastCalledWith('Wrong email address');

    component.instance().handleTextFieldChange({ target: { id: '', value: 'd@d.c' } }); // For switch default
  });

  test('handleRegisterBtnClick', () => {
    const state = { username: 'username', password: 'password', email: 'email' };
    const component = getShallowComponent();
    component.instance().handleRegisterBtnClick();
    component.setState(state);
    expect(component.state('isReady')).toBe(false);
    expect(component.state('isWaiting')).toBe(true);
    expect(defaultProps.registerUser).toHaveBeenCalledTimes(1);
  });

  test('onTogglePanels', () => {
    const component = getShallowComponent();
    component.find('Button').at(1).simulate('click');
    expect(defaultProps.onTogglePanels).toHaveBeenCalledTimes(1);
  });

  test('Snapshot username state equal empty', () => expect(renderer.create(<RegisterPanel {...defaultProps} />).toJSON()).toMatchSnapshot());
  test('Snapshot username state not equal empty', () => {
    const component = getShallowComponent();
    component.setState({ username: 'username' });
    expect(renderer.create(component).toJSON()).toMatchSnapshot();
  });
});
