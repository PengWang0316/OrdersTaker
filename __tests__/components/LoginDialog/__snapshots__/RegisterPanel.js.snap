// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterPanel Snapshot username state equal empty 1`] = `
Array [
  <Typography
    className="title"
    color="primary"
    data-testid="registerPanelTitle"
  >
    Register a new user
  </Typography>,
  <form
    autoComplete="off"
    noValidate={true}
  >
    <div>
      <Tooltip
        id="tooltipUsername"
        title="Characters, number, _ and @"
      >
        <TextField
          autoFocus={true}
          disabled={false}
          error={false}
          fullWidth={true}
          helperText={null}
          id="username"
          label="Username"
          margin="normal"
          onChange={[Function]}
          type="email"
          value=""
        />
      </Tooltip>
    </div>
    <div>
      <TextField
        disabled={false}
        fullWidth={true}
        id="password"
        label="Password"
        margin="normal"
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <Tooltip
      id="tooltipRpPw"
      title="Please type in your password again"
    >
      <TextField
        disabled={false}
        error={false}
        fullWidth={true}
        helperText={null}
        id="repeatPassword"
        label="Repeat Password"
        margin="normal"
        onChange={[Function]}
        type="password"
        value=""
      />
    </Tooltip>
    <div>
      <TextField
        disabled={false}
        error={false}
        fullWidth={true}
        helperText={null}
        id="email"
        label="Email"
        margin="normal"
        onChange={[Function]}
        type="email"
        value=""
      />
    </div>
    <Tooltip
      id="tooltipRegisterBtn"
      placement="bottom-end"
      title="Please fill out all field above"
    >
      <div
        className="registerBtnDiv"
      >
        <Button
          aria-label="Register"
          color="primary"
          disabled={true}
          onClick={[Function]}
          size="small"
          variant="contained"
        >
          Register
        </Button>
      </div>
    </Tooltip>
    <div>
      <Button
        className="switchBtn"
        color="primary"
        data-testid="switchLoginButton"
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          }
        }
      >
        Have an account already? Go to login
      </Button>
    </div>
  </form>,
]
`;

exports[`RegisterPanel Snapshot username state not equal empty 1`] = `
Array [
  <Typography
    className="title"
    color="primary"
    data-testid="registerPanelTitle"
  >
    Register a new user
  </Typography>,
  <form
    autoComplete="off"
    noValidate={true}
  >
    <div>
      <Tooltip
        id="tooltipUsername"
        title="Characters, number, _ and @"
      >
        <TextField
          autoFocus={true}
          disabled={false}
          error={false}
          fullWidth={true}
          helperText={null}
          id="username"
          label="Username"
          margin="normal"
          onChange={[Function]}
          type="email"
          value="username"
        />
      </Tooltip>
      <CheckingIndicator
        isAvaliable={false}
        isChecking={false}
      />
    </div>
    <div>
      <TextField
        disabled={false}
        fullWidth={true}
        id="password"
        label="Password"
        margin="normal"
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <Tooltip
      id="tooltipRpPw"
      title="Please type in your password again"
    >
      <TextField
        disabled={false}
        error={false}
        fullWidth={true}
        helperText={null}
        id="repeatPassword"
        label="Repeat Password"
        margin="normal"
        onChange={[Function]}
        type="password"
        value=""
      />
    </Tooltip>
    <div>
      <TextField
        disabled={false}
        error={false}
        fullWidth={true}
        helperText={null}
        id="email"
        label="Email"
        margin="normal"
        onChange={[Function]}
        type="email"
        value=""
      />
    </div>
    <Tooltip
      id="tooltipRegisterBtn"
      placement="bottom-end"
      title="Please fill out all field above"
    >
      <div
        className="registerBtnDiv"
      >
        <Button
          aria-label="Register"
          color="primary"
          disabled={true}
          onClick={[Function]}
          size="small"
          variant="contained"
        >
          Register
        </Button>
      </div>
    </Tooltip>
    <div>
      <Button
        className="switchBtn"
        color="primary"
        data-testid="switchLoginButton"
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "isThrow": false,
                "value": undefined,
              },
            ],
          }
        }
      >
        Have an account already? Go to login
      </Button>
    </div>
  </form>,
]
`;
